// vite-plugin-svg-sprite/cjs/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
const crypto_1 = __importDefault(require("crypto"));
const micromatch_1 = __importDefault(require("micromatch"));
const svg_baker_1 = __importDefault(require("svg-baker"));
const svg_parser_1 = require("svg-parser");
const svgo_1 = require("svgo");
const { stringify } = JSON;
exports.default = (options) => {
    var _a;
    const svgCompiler = new svg_baker_1.default();
    const match = (_a = options === null || options === void 0 ? void 0 : options.include) !== null && _a !== void 0 ? _a : '**.svg';
    const svgoOptions = options === null || options === void 0 ? void 0 : options.svgo;
    const plugin = {
        name: 'svg-sprite',
        async transform(src, filepath) {
            var _a, _b, _c;
            if (!micromatch_1.default.isMatch(filepath, match, {
                dot: true,
            })) {
                return undefined;
            }
            let code = await fs_1.default.promises.readFile(filepath, 'utf-8');
            const root = (0, svg_parser_1.parse)(code);
            let topLevelAttributes = {};
            if (((_a = root.children[0]) === null || _a === void 0 ? void 0 : _a.type) === 'element') {
                topLevelAttributes = (_b = root.children[0].properties) !== null && _b !== void 0 ? _b : {};
            }
            const { name } = path_1.default.parse(filepath);
            if (svgoOptions !== false) {
                const result = ((0, svgo_1.optimize)(code, svgoOptions === true ? undefined : svgoOptions));
                code = result.data;
            }
            let id = name;
            if (options === null || options === void 0 ? void 0 : options.symbolId) {
                id = options.symbolId;
                if (id.includes('[hash]')) {
                    const hash = crypto_1.default.createHash('sha256');
                    hash.update(code);
                    id = id.replace(/\[hash\]/g, hash.digest('hex').slice(0, 6));
                }
                id = id.replace(/\[name\]/g, name);
            }
            const symbol = await svgCompiler.addSymbol({
                id,
                content: code,
                path: filepath,
            });
            const codeToReturn = `
        import addSymbol from 'vite-plugin-svg-sprite/runtime';
        addSymbol(${stringify(symbol.render())}, ${stringify(id)});
        export default ${stringify(id)};
        export const size = { width: ${stringify(topLevelAttributes.width)}, height: ${stringify(topLevelAttributes.height)} };
      `;
            return {
                code: codeToReturn,
                moduleSideEffects: (_c = options === null || options === void 0 ? void 0 : options.moduleSideEffects) !== null && _c !== void 0 ? _c : true,
                map: { mappings: '' },
            };
        },
    };
    return plugin;
};
